packages/vm/src/context.rs

```diff
@@ -62,229 +58,243 @@ impl GasState {
+
+    pub fn set_storage_readonly(&self, new_value: bool) {
+        self.with_context_data_mut(|context_data| {
+            context_data.storage_readonly = new_value;
+        })
+    }
+
+    pub fn memory(&self) -> Memory {
+        self.with_context_data(|context| {
+            let ptr = context
+                .wasmer_instance
+                .expect("Wasmer instance is not set. This is a bug.");
+            let instance = unsafe { ptr.as_ref() };
+            let mut memories: Vec<Memory> = instance
+                .exports
+                .iter()
+                .memories()
+                .map(|pair| pair.1.clone())
+                .collect();
+            memories.pop().unwrap()
+        })
     }
 }
 
-/// Creates a back reference from a contact to its partent instance
-pub fn set_wasmer_instance<S: Storage, Q: Querier>(
-    ctx: &mut Ctx,
+pub struct ContextData<S: Storage, Q: Querier> {
+    gas_state: GasState,
+    storage: Option<S>,
+    storage_readonly: bool,
+    querier: Option<Q>,
+    /// A non-owning link to the wasmer instance
     wasmer_instance: Option<NonNull<WasmerInstance>>,
-) {
-    let context_data = ctx.data as *mut ContextData<S, Q>;
-    unsafe {
-        (*context_data).wasmer_instance = wasmer_instance;
+}
+
+impl<S: Storage, Q: Querier> ContextData<S, Q> {
+    pub fn new(gas_limit: u64) -> Self {
+        ContextData::<S, Q> {
+            gas_state: GasState::with_limit(gas_limit),
+            storage: None,
+            storage_readonly: true,
+            querier: None,
+            wasmer_instance: None,
+        }
     }
 }
 
 /// Returns the original storage and querier as owned instances, and closes any remaining
 /// iterators. This is meant to be called when recycling the instance.
 pub(crate) fn move_out_of_context<S: Storage, Q: Querier>(
-    source: &mut Ctx,
+    env: &Env<S, Q>,
 ) -> (Option<S>, Option<Q>) {
-    let b = get_context_data_mut::<S, Q>(source);
-    (b.storage.take(), b.querier.take())
+    env.with_context_data_mut(|context_data| {
+        (context_data.storage.take(), context_data.querier.take())
+    })
 }
 
-/// Moves owned instances of storage and querier into the context.
+/// Moves owned instances of storage and querier into the env.
 /// Should be followed by exactly one call to move_out_of_context when the instance is finished.
-pub(crate) fn move_into_context<S: Storage, Q: Querier>(target: &mut Ctx, storage: S, querier: Q) {
-    let b = get_context_data_mut::<S, Q>(target);
-    b.storage = Some(storage);
-    b.querier = Some(querier);
-}
-
-pub fn get_gas_state_mut<'a, 'b, S: Storage + 'b, Q: Querier + 'b>(
-    ctx: &'a mut Ctx,
-) -> &'b mut GasState {
-    &mut get_context_data_mut::<S, Q>(ctx).gas_state
-}
-
-pub fn get_gas_state<'a, 'b, S: Storage + 'b, Q: Querier + 'b>(ctx: &'a Ctx) -> &'b GasState {
-    &get_context_data::<S, Q>(ctx).gas_state
+pub(crate) fn move_into_context<S: Storage, Q: Querier>(env: &Env<S, Q>, storage: S, querier: Q) {
```

Comment:
Okay, so we still have the same format here.

No personal problem with this, but I remember it was one thing that was a criticism in the old structure. I didn't see a way to remove it either.

Reply:
The general issue still exists: you need to create the host functions before you have the instance, but want to access the instance from the host functions.

The access is much safer now, but the core problem remains. Wasmer recently added a common API for it, which creates a lazy loading machanism. Not much different from what we do here, but a common way for everyone. Will add at some point.